inbm/dispatcher-agent/dispatcher/sota/oras_util.py

```diff
@@ -0,0 +1,150 @@
+"""
+    ORAS tool will be called by dispatcher to perform the image downloading in TiberOS.
+
+    Copyright (C) 2017-2024 Intel Corporation
+    SPDX-License-Identifier: Apache-2.0
+"""
+import logging
+import os
+import requests
+import json
+from urllib.parse import urlsplit
+from typing import Optional
+from dispatcher.dispatcher_exception import DispatcherException
+from inbm_common_lib.shell_runner import PseudoShellRunner
+from inbm_common_lib.utility import CanonicalUri
+from dispatcher.packagemanager.local_repo import DirectoryRepo
+from dispatcher.packagemanager.package_manager import verify_source
+from ..packagemanager.irepo import IRepo
+from ..dispatcher_broker import DispatcherBroker
+from ..constants import CACHE
+
+
+logger = logging.getLogger(__name__)
+
+
+def oras_download(dispatcher_broker: DispatcherBroker, uri: CanonicalUri,
+             repo: IRepo, username: Optional[str], password: str, umask: int) -> None:
+    """Downloads files and places capsule file in path mentioned by manifest file.
+
+    @param dispatcher_broker: DispatcherBroker object used to communicate with other INBM services
+    @param uri: URI of the source location
+    @param repo: repository for holding the download
+    @param username: username to use for download
+    @param password: password to use for download
+    @param umask: file permission mask
+    @raises DispatcherException: any exception
+    """
+    dispatcher_broker.telemetry(f'Package to be fetched from {uri.value}')
+    dispatcher_broker.telemetry(
+        'Checking authenticity of package by checking signature and source')
+
+    if not isinstance(uri, CanonicalUri):
+        raise DispatcherException("Internal error: uri improperly passed to download function")
+
+    """
+    In case of uri.value = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    source = https://registry-rs.internal.ledgepark.intel.com/one-intel-edge
+    registry_server = registry-rs.internal.ledgepark.intel.com
+    image = tiberos
+    image_tag = latest
+    image_full_path = registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest
+    repository_name = one-intel-edge
+    registry_manifest = https://registry-rs.internal.ledgepark.intel.com/v2/one-intel-edge/tiberos/manifest/latest
+    """
+    try:
+        source = uri.value[:-(len(uri.value.split('/')[-1]) + 1)]
+        registry_server = uri.value.split("/")[2]
+        image = os.path.basename(urlsplit(uri.value).path).split(":")[0]
+        image_tag = uri.value.split(":")[-1]
+        repository_name = "/".join(uri.value.split("/")[3:-1])
+        image_full_path = f"{registry_server}/{repository_name}/{image}:{image_tag}"
+        registry_manifest = f"https://{registry_server}/v2/{repository_name}/{image}/manifests/{image_tag}"
+    except IndexError as err:
+        logger.error(f"IndexError occurs with uri {uri.value}: {err}")
```

Comment:
error could be confusing as user will only see `IndexError` and not know what part of the parsing failed

Reply:
I can only think of adding the try and except in each line of the parsing so that the user will know which part of the parsing failed. Is there a more effective solution?

Reply:
@yengliong93

I ran the question through IGPT and I got this suggestion:

```python
@dataclass
class ParsedURI:
    source: str
    registry_server: str
    image: str
    image_tag: str
    image_full_path: str
    repository_name: str
    registry_manifest: str

def parse_uri(uri: CanonicalUri) -> ParsedURI:
    """
    Parse a CanonicalUri and extract its components for a container image URI.
    
    Args:
        uri (CanonicalUri): A URI object with a value containing a full image URI.
        
        Example URI: https://registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest

    Returns:
        ParsedURI: A data structure containing the following fields:
        
        - source: The base URL of the registry, including the repository (e.g., 'https://registry-rs.internal.ledgepark.intel.com/one-intel-edge')
        - registry_server: The server hosting the registry (e.g., 'registry-rs.internal.ledgepark.intel.com')
        - image: The image name (e.g., 'tiberos')
        - image_tag: The image tag (e.g., 'latest')
        - image_full_path: The full path to the image with tag (e.g., 'registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest')
        - repository_name: The repository name (e.g., 'one-intel-edge')
        - registry_manifest: The URL to the image's manifest (e.g., 'https://registry-rs.internal.ledgepark.intel.com/v2/one-intel-edge/tiberos/manifest/latest')
    
    Raises:
        ValueError: If the URI path does not contain enough components to extract the necessary information.
    
    Example:
        uri = CanonicalUri('https://registry-rs.internal.ledgepark.intel.com/one-intel-edge/tiberos:latest')
        parsed_uri = parse_uri(uri)
        print(parsed_uri)
    """
    
    # Parse the URI
    parsed = urlparse(uri.value)
    registry_server = parsed.netloc

    # Split the path and remove empty segments
    path_parts = [part for part in parsed.path.strip('/').split('/') if part]
    if len(path_parts) < 2:
        raise ValueError("URI path does not contain enough components.")

    # Extract repository name and image details
    repository_name = path_parts[0]
    image_with_tag = path_parts[1]

    # Extract image and tag (default tag is 'latest' if not provided)
    if ':' in image_with_tag:
        image, image_tag = image_with_tag.split(':', 1)
    else:
        image = image_with_tag
        image_tag = 'latest'

    # Construct the required fields
    source = f"{parsed.scheme}://{registry_server}/{repository_name}"
    image_full_path = f"{registry_server}/{repository_name}/{image_with_tag}"
    registry_manifest = (
        f"{parsed.scheme}://{registry_server}/v2/"
        f"{repository_name}/{image}/manifest/{image_tag}"
    )

    # Return the populated ParsedURI data structure
    return ParsedURI(
        source=source,
        registry_server=registry_server,
        image=image,
        image_tag=image_tag,
        image_full_path=image_full_path,
        repository_name=repository_name,
        registry_manifest=registry_manifest,
    )
 ```

Reply:
Added ParsedURI dataclass and updated the function.